scheds/rust/scx_lavd/src/main.rs

```diff
@@ -0,0 +1,308 @@
+        let tc = mt.taskc;
+
+        // No idea how to print other types than LAVD_MSG_TASKC
+        if mt.hdr.kind != LAVD_MSG_TASKC {
+            return 0;
+        }
+
+        // Print a message from the BPF scheduler
+        let mseq = Scheduler::get_msg_seq_id();
+
+        if mseq % 32 == 1 {
+            info!("| {:9} | {:8} | {:17} \
+                   | {:4} | {:9} | {:9} \
+                   | {:10} | {:9} | {:8} \
+                   | {:12} | {:7} | {:9} \
+                   | {:9} | {:9} | {:9} \
+                   | {:9} | {:8} |",
+                  "mseq", "pid", "comm",
+                  "cpu", "vddln_ns", "elglty_ns",
+                  "slice_ns", "grdy_rt", "lat_prio",
+                  "static_prio", "lat_bst", "slice_bst",
+                  "run_freq", "run_tm_ns", "wait_freq",
+                  "wake_freq", "cpu_util");
+        }
+
+        let c_tx_cm: *const c_char = (&tx.comm as *const [i8;17]) as *const i8;
+        let c_tx_cm_str: &CStr = unsafe { CStr::from_ptr(c_tx_cm) };
+        let tx_comm : &str = c_tx_cm_str.to_str().unwrap();
+
+        info!("| {:9} | {:8} | {:17} \
+               | {:4} | {:9} | {:9} \
+               | {:10} | {:9} | {:8} \
+               | {:12} | {:7} | {:9} \
+               | {:9} | {:9} | {:9} \
+               | {:9} | {:8} | ",
+              mseq, tx.pid, tx_comm,
+              tx.cpu_id, tc.vdeadline_delta_ns, tc.eligible_delta_ns,
+              tc.slice_ns, tc.greedy_ratio, tc.lat_prio,
+              tx.static_prio, tc.lat_boost_prio, tc.slice_boost_prio,
+              tc.run_freq, tc.run_time_ns, tc.wait_freq,
+              tc.wake_freq, tx.cpu_util);
+
+        0
+    }
+
+    fn prep_introspec(&mut self) -> u64 {
+        let mut interval_ms = 1000;
+
+        if self.intrspc.cmd == LAVD_CMD_SCHED_N && 
+            self.intrspc.arg > self.nr_cpus_onln {
+		// More samples, shorter sampling interval.
+                let f = self.intrspc.arg / self.nr_cpus_onln * 2;
+                interval_ms /= f;
+        }
+	self.intrspc.requested = true as u8;
+
+	self.skel.bss_mut().intrspc.cmd = self.intrspc.cmd;
+	self.skel.bss_mut().intrspc.arg = self.intrspc.arg;
+	self.skel.bss_mut().intrspc.requested = self.intrspc.requested;
+
+        interval_ms
+    }
+
+    fn cleanup_introspec(&mut self) {
+        // If not yet requested, do nothing.
+        if self.intrspc.requested == false as u8 {
+            return;
+        }
+
+        // Once dumped, it is done.
+        if self.intrspc.cmd == LAVD_CMD_DUMP {
+            self.intrspc.cmd = LAVD_CMD_NOP;
+	}
+    }
+
+    fn running(&mut self) -> bool {
+        RUNNING.load(Ordering::Relaxed) && !uei_exited!(&self.skel.bss().uei)
+    }
+
+    fn run(&mut self) -> Result<()> {
+        while self.running() {
+            let interval_ms = self.prep_introspec();
+            std::thread::sleep(Duration::from_millis(interval_ms));
+            self.rb_mgr.poll(Duration::from_millis(100)).unwrap();
```

Comment:
Is there any reason to have this `std::thread::sleep()` here? Can we just do:
```
        self.rb_mgr.poll(Duration::from_millis(interval_ms)).unwrap();
```

Reply:
That's because it is assumed self.prep_introspec() is called every interval_ms micro-seconds. So self.prep_introspec() resets the introspection command every call.